add maps feature to site

poncem91 opened the pull request:
Adds a map that allows logged in users to add markers to share where they are visiting from. The markers are stored permanently with datastore. When the site finishes loading, the map will load with all previously added markers and will allow logged in users to add new markers or delete markers that they themselves have added in the past.

poncem91 committed:
add basic map using google maps api

Site loads a map under the comment history div once the page is loaded.

poncem91 committed:
adds feature that lets users add markers to map

Commit allows users to click on the map and add markers permanently in the map by storing their coordinates in datastore.

poncem91 committed:
Merge branch 'Mafe_Authentication' into Mafe_Maps

merge changes from base branch

poncem91 committed:
store user's userId with every marker permanently in datastore

Commit stores the id of users who make each marker to identify who is the creator of each marker.

poncem91 committed:
add delete marker functionality

poncem91 committed:
change map functionality so only logged users can add markers

poncem91 committed:
change wording for login message

poncem91 committed:
add location address to infowindows when markers are clicked

poncem91 committed:
add address to marker infowindow and allow user to close infowindows by clicking on map

When users click on existing markers, the marker's infowindow will show the location's address. Additionally, users are now able to close out all infowindows by simply clicking anywhere in the map. Consequently, in order to add a new marker by clicking on the map all infowindows must first be closed.

peap commented on portfolio/.gitignore:
```diff
@@ -0,0 +1 @@
+config.js
```
When adding an uncommitted config file to a repo, I like to leave a template or instructions in a README about what is expected to exist in that config file, so that other developers can replicate it. It would also be helpful for this .gitignore file to list the full path to the file we're ignoring, which I suspect is src/main/webapp/config.js, right?

From looking at the code, I suspect the file needs to contain something like:

```
var config = {MAPS_KEY: 'my-api-key-for-google-maps'};
```

poncem91 replied:
Sounds good!

peap commented on portfolio/src/main/java/com/google/sps/servlets/MarkersServlet.java:
```diff
@@ -0,0 +1,127 @@
+        query.setFilter(queryFilter);
+
+        PreparedQuery results = datastore.prepare(query);
+        Entity entity = results.asSingleEntity();
```
Is this one Entity actually a list of Marker entities?

poncem91 replied:
This should be a single marker, but now that you mention it there might exist a chance of the same user creating two exact markers so this would result in multiple entities within this query. I'll change this so it uses the unique id instead of lat and lng to avoid this scenario!

peap replied:
Sounds good!

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -203,4 +201,252 @@ function getFilter() {
+        const deleteButton = document.createElement('button');
+        deleteButton.innerText = "Delete";
+        deleteButton.onclick = () => {
+            markersMap.get(marker.get('id')).setMap(null);
+            markersMap.delete(marker.get('id'));
```
Maybe `deleteMarker` ought to handle removing the marker from the map, as well as from the database.

poncem91 replied:
That would work better, I'll move things around to achieve all of this within deleteMarker(). Thanks!

poncem91 committed:
add details to README.md, change deleteMarker implementation, add java helper delete methods

Details regarding what the config.js file should include were added to the README.md file, so others know to include their own Google Maps API key in there. The deleteMarker() function now handles both the front-end removal of markers and the back-end. The back-end removal of markers is now implemented using the marker's id instead of its lat and lng values, to avoid situations where querying the marker to delete results in multiple markers when there exists multiple markers with the same coordinates. A helper class and helper methods were added in java to assist in the deletion of comments and markers, avoiding some code repetition and clutter.

poncem91 committed:
add change to README.md file

poncem91 committed:
add delete all your markers custom control to map

Commit adds a Google Maps custom control that allows logged in users to delete all of their previously added markers from the map.

poncem91 committed:
add comment for loading the delete all markers custom control

peap commented on portfolio/src/main/java/com/google/sps/data/Entities.java:
```diff
@@ -0,0 +1,72 @@
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Key entityKey = KeyFactory.createKey(kind, id);
+
+        Query query = new Query(kind);
+
+        UserService userService = UserServiceFactory.getUserService();
+        String userId = userService.getCurrentUser().getUserId();
+
+        Query.CompositeFilter queryFilter = new Query.CompositeFilter(Query.CompositeFilterOperator.AND, Arrays
+            .asList(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userId), new Query
+            .FilterPredicate("__key__", Query.FilterOperator.EQUAL, entityKey)));
+
+        query.setFilter(queryFilter);
+        PreparedQuery results = datastore.prepare(query);
+        Entity entity = results.asSingleEntity();
```
Filtering on the special `__key__` entry *and* other properties looks a bit strange to me. I would recommend leaving a comment here about not wanting users to be able to delete other users' entities, which is why you have this filtering on both `userId` and the key.

poncem91 replied:
Got it, I'll add some comments to explain that! And yes, I'm filtering both because I want to make sure only the logged user can delete its own comments in the back end.

peap commented on portfolio/src/main/java/com/google/sps/data/Entities.java:
```diff
@@ -0,0 +1,72 @@
+        ArrayList<Long> deletedIds = new ArrayList<>();
+
+        for (Entity entity : results.asIterable()) {
+            datastore.delete(entity.getKey());
+            long id = entity.getKey().getId();
```
It's apparently working, but I would store the ID of the entity before I deleted it, personally. I think that some database frameworks you use in the future might update your local references to the data in the database after you delete it, in which case `entity.getKey().getId()` might not work after the item is deleted.

poncem91 replied:
That makes perfect sense, I should have thought of that beforehand. Thanks for catching that!

poncem91 committed:
add comment on Entities.java and get deleted entityId before deleting entity

poncem91 committed:
improve javasript styling

Stop using var and only use const and let instead. Be consistent with quotation use by making all of them double quotes instead of single.

poncem91 committed:
add git code format maven plugin

peap commented on portfolio/src/main/java/com/google/sps/servlets/MarkersServlet.java:
```diff
@@ -35,83 +35,83 @@
+
+      if (id > 0) {
+        Entities.deleteSingle(id, "Marker");
+        response.setContentType("text/html;");
```
We probably ought to have a given endpoint return the same type of response in all cases; so, always JSON, always HTML, etc.

poncem91 replied:
Sounds good, I'll change this to respond with JSON as well, thanks!

Would this apply to different HTTP method requests as well? For example in DataServlet.java the type of response for doDelete is text/html but for doGet is JSON - should I also have these return the same response type as well or does it only have to be the same response type within the same request method?

peap replied:
As discussed in person, I just mean keeping it consistent with a (URL, method) pair. So, DELETE /markers should always return JSON, whereas it could make sense for POST /markers to always return HTML. So yeah, just keep the response type the same with each servlet handler method.

poncem91 committed:
change all response types to JSON in MarkersServlet doDelete

Changes the response type of deleting single markers to JSON to keep the response type consistent within the /markers endpoint.

poncem91 committed:
Merge branch 'master' into Mafe_Maps

merge changes to base branch
